File: priv/src/exdl_mouse.c
/*
 *  Copyright (c) 2001 Dan Gudmundsson
 *
 *  See the file "license.terms" for information on usage and redistribution
 *  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 *     $Id: exdl_mouse.c,v 1.2 2005/06/29 14:45:38 dgud Exp $
 */

#include "exdl.h"

EXDL_API(emou_getMouseState)
{
  /* Uint8 SDL_GetMouseState(int* x, int* y) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  FRAME_FMT_BEGIN()
    FRAME_FMT_DATA(Uint8, status)
    FRAME_FMT_DATA(int,   x)
    FRAME_FMT_DATA(int,   y)
  FRAME_FMT_END()

  FRAME_SET_DATA(status) =
//  CATCH_RESULT(Uint8)

    SDL_GetMouseState
    ( FRAME_GET_ADDR(x)
    , FRAME_GET_ADDR(y)
    );

  THROW_FRAME()
}

EXDL_API(emou_getRelativeMouseState)
{
  /* Uint8 SDL_GetRelativeMouseState(int* x, int* y) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  FRAME_FMT_BEGIN()
    FRAME_FMT_DATA(Uint8, status)
    FRAME_FMT_DATA(int,   x)
    FRAME_FMT_DATA(int,   y)
  FRAME_FMT_END()

  FRAME_SET_DATA(status) =
//  CATCH_RESULT(Uint8)

    SDL_GetRelativeMouseState
    ( FRAME_GET_ADDR(x)
    , FRAME_GET_ADDR(y)
    );

  THROW_FRAME()
}

EXDL_API(emou_warpMouse)
{
  /* void SDL_WarpMouse(Uint16 x, Uint16 y) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(Uint16, x)
    STACK_FMT_DATA(Uint16, y)
  STACK_FMT_END()

    SDL_WarpMouse
    ( STACK_GET_DATA(x)
    , STACK_GET_DATA(y)
    );
}

EXDL_API(emou_createCursor)
{
  /* SDL_Cursor* SDL_CreateCursor(Uint8* data, Uint8* mask, int w, int h, int hot_x, int hot_y) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int,   w)
    STACK_FMT_DATA(int,   h)
    STACK_FMT_DATA(int,   hotx)
    STACK_FMT_DATA(int,   hoty)
    STACK_FMT_DATA(int,   ds)
    STACK_FMT_DATA(Uint8, data[1]) /* fake size */
    STACK_FMT_DATA(Uint8, mask[1])
  STACK_FMT_END()

  CATCH_RESULT(SDL_Cursor*)

    SDL_CreateCursor
    ( STACK_GET_ADDR(data)
    , STACK_GET_ADDR(data)
    + STACK_GET_DATA(ds) /* mask */
    , STACK_GET_DATA(w)
    , STACK_GET_DATA(h)
    , STACK_GET_DATA(hotx)
    , STACK_GET_DATA(hoty)
    );

  THROW_POINTER()
}

EXDL_API(emou_setCursor)
{
  /* void SDL_SetCursor(SDL_Cursor* cursor) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Cursor, cursor[])
  STACK_FMT_END()

    SDL_SetCursor
    ( STACK_GET_ADDR(cursor)
    );
}

EXDL_API(emou_getCursor)
{
  /* SDL_Cursor* SDL_GetCursor(void) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  CATCH_RESULT(SDL_Cursor*)

    SDL_GetCursor
    (
    );

  THROW_POINTER()
}

EXDL_API(emou_freeCursor)
{
  /* void SDL_FreeCursor(SDL_Cursor* cursor) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Cursor, cursor[])
  STACK_FMT_END()

    SDL_FreeCursor
    ( STACK_GET_ADDR(cursor)
    );
}

EXDL_API(emou_showCursor)
{
  /* int SDL_ShowCursor(int toggle) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int, toggle)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_ShowCursor
    ( STACK_GET_DATA(toggle)
    );

  THROW_RESULT(int)
}

File: src/exdl_mouse.erl
%%
%%  Copyright (c) 2001 Dan Gudmundsson
%%
%%  See the file "license.terms" for information on usage and redistribution
%%  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
%% 
%%     $Id: exdl_mouse.erl,v 1.2 2005/06/29 14:45:39 dgud Exp $
%%

-module(exdl_mouse).

%% 1- EXPORT

%-compile(export_all).

-export( [ createCursor/6
         , freeCursor/1
         , getCursor/0
         , getMouseState/0
         , getRelativeMouseState/0
         , setCursor/1
         , showCursor/1
         , warpMouse/2
         ]
       ).

%% 2- INCLUDE

-include("exdl.hrl").

-include("exdl_util.hrl").

-include("exdl_mouse.hrl").
%-include("exdl_mouse_fp.hrl").

%% 3- IMPORT

-import( exdl_driver
       , [ call/2
         , cast/2
         ]
       ).

%% 4- FUNCTIONS

%% Func:       getMouseState
%% Args:       none
%% Returns:    { MouseState, X, Y }
%% C-API func: Uint8 SDL_GetMouseState(int* x, int* y)
%% Desc:

getMouseState()
 -> <<State:8
    , X:16
    , Y:16
    >>
    = call
      ( ?SDL_GetMouseState
      , [
        ]
      )
  , { State
    , X
    , Y
    }
.

%% Func:       getRelativeMouseState
%% Args:       none
%% Returns:    { MouseState, X, Y }
%% C-API func: Uint8 SDL_GetRelativeMouseState(int* x, int* y);
%% Desc:

getRelativeMouseState()
 -> <<State:8
    , X:16
    , Y:16
    >>
    = call
      ( ?SDL_GetRelativeMouseState
      , [
        ]
      )
  , { State
    , X
    , Y
    }
.

%% Func:       warpMouse
%% Args:       X, Y
%% Returns:    ok 
%% C-API func: void SDL_WarpMouse(Uint16 x, Uint16 y)

warpMouse(X, Y)
 -> cast
    ( ?SDL_WarpMouse
    , <<X:?Uint16
      , Y:?Uint16
      >>
    )
.

%% Func:        createCursor
%% Args:        Data (Binary), Mask (Binary), W, H, HotX, HotY
%% Returns:     CursorRef
%% C-API func:  SDL_Cursor* SDL_CreateCursor(Uint8* data, Uint8* mask, int w, int h, int hot_x, int hot_y);
%% Desc:        Data & Mask must be less than 32*32 bytes.

createCursor(Data, Mask, W, H, HotX, HotY)
  when is_binary(Data), is_binary(Mask)
 -> case
      call
      ( ?SDL_CreateCursor
      , <<W:16/native
        , H:16/native
        , HotX:16/native
        , HotY:16/native
        , (size(Data)):16/native
        , Data/binary
        , Mask/binary
        >>
      )
    of
      <<0:64
      >>
        -> exit
           ( { createCursor
             , returned_null
             }
           )
    ; <<Ptr:?_PTR
      >>
        -> { cursorp
           , Ptr
           }
    end
.

%% Func:        setCursor
%% Args:        CursorRef
%% Returns:     ok
%% C-API func:  void SDL_SetCursor(SDL_Cursor* cursor);

setCursor({cursorp, Ref})
 -> cast
    ( ?SDL_SetCursor
    , <<Ref:?_PTR
      >>
    )
.

%% Func:        getCursor
%% Args:        none
%% Returns:     A cursorRef
%% C-API func:  void SDL_SetCursor(SDL_Cursor* cursor);

getCursor()
 -> case
      call
      ( ?SDL_GetCursor
      , [
        ]
      )
    of
      <<0:?_PTR
      >>
        -> exit
           ( { getCursor
             , returned_null
             }
           )
    ; <<Ptr:?_PTR
      >>
        -> { cursorp
           , Ptr
           }
    end
.

%% Func:        freeCursor
%% Args:        CursorRef
%% Returns:     ok
%% C-API func:  void SDL_FreeCursor(SDL_Cursor* cursor);

freeCursor({cursorp,Ref})
 -> cast
    ( ?SDL_FreeCursor
    , <<Ref:?_PTR
      >>
    )
.

%% Func:        showCursor
%% Args:        true | false
%% Returns:     true | false (if cursor was displayed before the call)
%% C-API func:  int SDL_ShowCursor(int toggle);

showCursor(Bool)
 -> B
    = case
        Bool
      of
        false
          -> 0
      ; true 
          -> 1
      end
  , <<Res:8
    >>
    = call
      ( ?SDL_ShowCursor
      , [ B
        ]
      )
  , Res =:= 1
.
